libyul/Object.cpp

```diff
@@ -61,13 +61,48 @@ string Object::toString(Dialect const* _dialect) const
 	return "object \"" + name.str() + "\" {\n" + indent(inner) + "\n}";
 }
 
-set<YulString> Object::dataNames() const
+set<YulString> Object::qualifiedDataNames() const
 {
-	set<YulString> names;
-	names.insert(name);
-	for (auto const& subObject: subIndexByName)
-		names.insert(subObject.first);
+	set<YulString> qualifiedNames = {name};
+	for (shared_ptr<ObjectNode> const& subObjectNode: subObjects)
+	{
+		qualifiedNames.insert(subObjectNode->name);
+		if (auto const* subObject = dynamic_cast<Object const*>(subObjectNode.get()))
+			for (YulString const& subSubObj: subObject->qualifiedDataNames())
+				qualifiedNames.insert(YulString{subObject->name.str() + "." + subSubObj.str()});
+	}
+
 	// The empty name is not valid
-	names.erase(YulString{});
-	return names;
+	qualifiedNames.erase(YulString{});
```

Comment:
I replaced it with an assertion, it was part of old code, so I just added it. But after another look at it, makes sense to remove and add assertion.

Reply:
It turns out that it was there because it really could happen in some paths of the code. I would leave it as is for now as it was there before.

Reply:
ok. But in that case I think it would be best to add a comment to the `name` member saying that it can be an empty string (and in what circumstances if you know that). I think it's not obvious.

Reply:
Hm, it happens when we have code like this (as example):
```
{
    {
        let a := 1
    }
    function f() {}
}
```

I will change a code a bit and add an assertion

Reply:
Oh, so that's when .yul file does not contain objects but just raw code.

Reply:
I think now it is fine. I add only name if it is not empty string.
if there are objects, name cannot be empty. But, assert at the end will make sure code will break if empty subObject name appears.

Reply:
So, since you have the `if` and the assert now, I think you can remove the `erase()`. It will never remove anything.

And I'd still add a docstring to `name` to inform about this corner case:
``` c++
/// Name of the object.
/// Can be empty since .yul files can also just contain code, without explicitly placing it in an object.
YulString name;
```